src/lib.rs

```diff
@@ -1013,6 +1107,220 @@ pub extern "C" fn next_list_stream_chunk(
+        let c_string = CString::new(format!("{}", error)).expect("should not have nulls");
+        self.response.error_message = c_string.into_raw();
+    }
+}
+
+impl GetStreamResponseGuard {
+    unsafe fn new(response_ptr: *mut GetStreamResponse, handle: *const c_void) -> GetStreamResponseGuard {
+        let response = unsafe { &mut (*response_ptr) };
+        response.result = CResult::Uninitialized;
+
+        GetStreamResponseGuard { response, handle }
+    }
+    fn success(self, stream: Box<GetStreamWrapper>, object_size: usize) {
+        self.response.result = CResult::Ok;
+        self.response.stream = Box::into_raw(stream);
+        self.response.object_size = object_size as u64;
+        self.response.error_message = std::ptr::null_mut();
+    }
+}
+
+impl Drop for GetStreamResponseGuard {
+    fn drop(&mut self) {
+        self.notify_on_drop()
+    }
+}
+
+unsafe impl Send for GetStreamResponseGuard {}
+
+#[no_mangle]
+pub extern "C" fn get_stream(
+    path: *const c_char,
+    size_hint: usize,
+    config: *const Config,
+    response: *mut GetStreamResponse,
+    handle: *const c_void
+) -> CResult {
+    let response = unsafe { GetStreamResponseGuard::new(response, handle) };
+    let path = unsafe { std::ffi::CStr::from_ptr(path) };
+    let path: Path = path.to_str().expect("invalid utf8").try_into().unwrap();
+    let config = unsafe { & (*config) };
+
+    match SQ.get() {
+        Some(sq) => {
+            match sq.try_send(Request::GetStream(path, size_hint, config, response)) {
+                Ok(_) => CResult::Ok,
+                Err(async_channel::TrySendError::Full(_)) => {
+                    CResult::Backoff
+                }
+                Err(async_channel::TrySendError::Closed(_)) => {
+                    CResult::Error
+                }
+            }
+        }
+        None => {
+            return CResult::Error;
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn read_get_stream(
+    stream: *mut GetStreamWrapper,
+    buffer: *mut u8,
+    size: usize,
+    amount: usize,
+    response: *mut ReadResponse,
+    handle: *const c_void
+) -> CResult {
+    let response = unsafe { ReadResponseGuard::new(response, handle) };
+    let mut slice = unsafe { std::slice::from_raw_parts_mut(buffer, size) };
+    let wrapper = match unsafe { stream.as_mut() } {
+        Some(w) => w,
+        None => {
+            tracing::error!("null stream pointer");
+            return CResult::Error;
+        }
+    };
+
+    match RT.get() {
+        Some(runtime) => {
+            runtime.spawn(async move {
+                let read_op = async {
+                    let amount_to_read = size.min(amount);
+                    let mut bytes_read = 0;
+                    while amount_to_read > bytes_read {
+                        let n = wrapper.reader.read_buf(&mut slice).await?;
+
+                        if n == 0 {
+                            let eof = amount_to_read != bytes_read;
```

Comment:
Shouldn't this just be eof = true?

Reply:
No, `read_buf` will also return 0 if the slice was completely filled, which is why we check if we read everything we requested and only if not we say it was an eof

Reply:
Taking another look, I think the while condition will never let us read 0 bytes and have `amount_to_read  == bytes_read` so eof will always be true as you said